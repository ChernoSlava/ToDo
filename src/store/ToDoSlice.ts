/* eslint-disable no-param-reassign */
import {
  ToDoStateType,
  EditAction,
  AddAction,
  IdAction,
  ToDoItemType,
} from '@types';
import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { LoadingState } from '@constants';

export const InitialToDo: ToDoStateType = {
  items: [],
  current: null,
  loading: LoadingState.Idle,
};

export const loadToDoList = createAsyncThunk(
  '@todo/list/load',
  async () => {
    const response = await fetch('/api/v1/todo');

    if (response.ok) {
      const result = await response.json();

      return result.list;
    }

    throw new Error('Error get todo list.')
  }
)

export const ToDoSlice = createSlice({
  name: 'ToDo_List',
  initialState: InitialToDo,
  reducers: {

    setLoadingState: (state: ToDoStateType, action: PayloadAction<LoadingState>) => {
      state.loading = action.payload;
    },

    edit: (state: ToDoStateType, action: EditAction) => {
      state.items = state.items.map(x => ({
        ...x,
        title: x.id === action.payload.id ? action.payload.title : x.title,
      }));
    },
    setCurrent: (state: ToDoStateType, action: PayloadAction<ToDoItemType>) => {
      state.current = action.payload;
    },
    remove: (state: ToDoStateType, action: IdAction) => {
      state.items = state.items.filter(x => x.id !== action.payload);
    },
    finish: (state: ToDoStateType, action: IdAction) => {
      state.items = state.items.map(x => ({
        ...x,
        isFinish: x.id === action.payload ? true : x.isFinish,
      }));
    },
    revert: (state: ToDoStateType, action: IdAction) => {
      state.items = state.items.map(x => ({
        ...x,
        isFinish: x.id === action.payload ? false : x.isFinish,
      }));
    },

    add: (state: ToDoStateType, action: AddAction) => {
      state.items.push(action.payload);
    },
    load: (state: ToDoStateType, action: PayloadAction<Array<ToDoItemType>>) => {
      state.items = action.payload;
    },
  },

  extraReducers: builder => {
    builder.addCase(loadToDoList.pending, state => {
      state.loading = LoadingState.Pending;
    });
    builder.addCase(loadToDoList.fulfilled, (state, action) => {
      state.loading = LoadingState.Success;
      state.items = action.payload
    });
    builder.addCase(loadToDoList.rejected, state => {
      state.loading = LoadingState.Fail;
    });
  }
});

export const { edit, remove, finish, revert, add, load, setLoadingState, setCurrent } =
  ToDoSlice.actions;

const { reducer: ToDoSliceReducer } = ToDoSlice;
export { ToDoSliceReducer };
