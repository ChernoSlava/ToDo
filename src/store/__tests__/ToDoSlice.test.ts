import { LoadingState } from "@constants";
import { addToDoItem, InitialToDo, loadToDoList, ToDoSliceReducer, editToDoItem } from "../ToDoSlice"
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { FetchMock } from 'jest-fetch-mock';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const propertyMatcherItem = {
  meta: {
    requestId: expect.any(String),
  },
};

const propertyErrorMatcherItem = {
  ...propertyMatcherItem,
  error: {
    stack: expect.any(String),
  },
};

describe('ToDoSlice test', () => {

  beforeEach(() => {
    fetchMock.doMock();
  });

  afterEach(() => {
    fetchMock.resetMocks();
  });
  //Tests for loadToDoList all
  describe('loadToDoList tests', () => {
    // Async loadToDoList function tests
    it('should success', async () => {
      const store = mockStore({ ...InitialToDo });

      (fetch as FetchMock).mockResponse(JSON.stringify({
        list: [{ id: 1 }]
      }), {
        status: 200,
      });

      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      await store.dispatch(await loadToDoList());

      expect(fetch).toBeCalledTimes(1);
      expect(fetch).toBeCalledWith('/api/v1/todo');

      const actions = store.getActions();

      expect({ actions }).toMatchSnapshot({
        actions: [
          propertyMatcherItem, propertyMatcherItem]
      });
    });

    it('should fail', async () => {
      const store = mockStore({ ...InitialToDo });

      (fetch as FetchMock).mockResponse(JSON.stringify({
        list: [{ id: 1 }]
      }), {
        status: 500,
      });

      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      await store.dispatch(await loadToDoList());

      expect(fetch).toBeCalledTimes(1);
      expect(fetch).toBeCalledWith('/api/v1/todo');

      const actions = store.getActions();

      expect({ actions }).toMatchSnapshot({
        actions: [
          propertyMatcherItem, propertyErrorMatcherItem]
      });
    });
    // extraReducer tests for loadToDoList
    it('should loadToDoList.pending change store correctly', () => {
      expect(ToDoSliceReducer(InitialToDo, { type: loadToDoList.pending.type }),
      ).toEqual({
        ...InitialToDo,
        loading: LoadingState.Pending,
      });
    });
    it('should loadToDoList.reject change store correctly', () => {
      expect(ToDoSliceReducer(InitialToDo, { type: loadToDoList.rejected.type }),
      ).toEqual({
        ...InitialToDo,
        loading: LoadingState.Fail,
      });
    });

    it('should loadToDoList.fulfild change store correctly', () => {
      expect(ToDoSliceReducer(InitialToDo, { type: loadToDoList.fulfilled.type, payload: [{}] }),
      ).toEqual({
        ...InitialToDo,
        loading: LoadingState.Success,
        items: [{}],
      });
    });
  });

  describe('addToDoList tests', () => {
    it('should success', async () => {
      const store = mockStore({ ...InitialToDo });

      (fetch as FetchMock).mockResponse(JSON.stringify({
        id: 100
      }), {
        status: 200,
      });

      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      await store.dispatch(await addToDoItem('item 1'));

      expect(fetch).toBeCalledTimes(1);
      expect(fetch).toBeCalledWith('/api/v1/todo',
        {
          "body": "{\"title\":\"item 1\"}",
          "headers": { "Content-Type": "application/json" },
          "method": "POST"
        });

      const actions = store.getActions();

      expect({ actions }).toMatchSnapshot({
        actions: [
          propertyMatcherItem, propertyMatcherItem]
      });
    });
    it('should fail', async () => {
      const store = mockStore({ ...InitialToDo });

      (fetch as FetchMock).mockResponse(JSON.stringify({
        id: 100
      }), {
        status: 500,
      });

      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      await store.dispatch(await addToDoItem('item 1'));

      expect(fetch).toBeCalledTimes(1);
      expect(fetch).toBeCalledWith('/api/v1/todo',
        {
          "body": "{\"title\":\"item 1\"}",
          "headers": { "Content-Type": "application/json" },
          "method": "POST"
        });

      const actions = store.getActions();

      expect({ actions }).toMatchSnapshot({
        actions: [
          propertyMatcherItem, propertyErrorMatcherItem]
      });
    });

    // extraReducer tests for addToDoItem
    it('should addToDoItem.pending change store correctly', () => {
      expect(ToDoSliceReducer(InitialToDo, { type: addToDoItem.pending.type }),
      ).toEqual({
        ...InitialToDo,
        loading: LoadingState.Pending,
      });
    });
    it('should addToDOList.reject change store correctly', () => {
      expect(ToDoSliceReducer(InitialToDo, { type: addToDoItem.rejected.type }),
      ).toEqual({
        ...InitialToDo,
        loading: LoadingState.Fail,
      });
    });

    it('should loadToDoList.fulfild change store correctly', () => {
      expect(ToDoSliceReducer(InitialToDo, {
        type: addToDoItem.fulfilled.type, payload: {
          title: 'as',
          isFinish: false,
          id: 'as',
        }
      }),
      ).toEqual({
        ...InitialToDo,
        loading: LoadingState.Success,
        items: [{
          title: 'as',
          isFinish: false,
          id: 'as',
        }],
      });
    });
  });

  // describe('editToDoItem tests', () => {
  //   it('should success', async () => {
  //     const store = mockStore({ ...InitialToDo });

  //     (fetch as FetchMock).mockResponse(JSON.stringify({
  //       title: 'a',
  //       id: 100,
  //     }), {
  //       status: 200,
  //     });

  //     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //     // @ts-ignore
  //     await store.dispatch(await editToDoItem({ title: 'a', id: 100, }));

  //     expect(fetch).toBeCalledTimes(1);
  //     expect(fetch).toBeCalledWith('/api/v1/todo/edit',
  //       {
  //         "body": "{\"title\":\"a\",\"id\":100}",
  //         "headers": { "Content-Type": "application/json" },
  //         "method": 'PATCH'
  //       });

  //     const actions = store.getActions();

  //     expect({ actions }).toMatchSnapshot({
  //       actions: [
  //         propertyMatcherItem, propertyMatcherItem]
  //     });
  //   });
  //   it('should fail', async () => {
  //     const store = mockStore({ ...InitialToDo });

  //     (fetch as FetchMock).mockResponse(JSON.stringify({
  //       id: 100
  //     }), {
  //       status: 500,
  //     });

  //     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //     // @ts-ignore
  //     await store.dispatch(await editToDoItem('item 1'));

  //     expect(fetch).toBeCalledTimes(1);
  //     expect(fetch).toBeCalledWith('/api/v1/todo/edit',
  //       {
  //         "body": "{\"title\":\"item 1\"}",
  //         "headers": { "Content-Type": "application/json" },
  //         "method": 'PATCH'
  //       });

  //     const actions = store.getActions();

  //     expect({ actions }).toMatchSnapshot({
  //       actions: [
  //         propertyMatcherItem, propertyErrorMatcherItem]
  //     });
  //   });
  // });

})