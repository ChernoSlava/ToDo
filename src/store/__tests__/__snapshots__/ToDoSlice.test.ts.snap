// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoSlice test addToDoList tests should fail 1`] = `
{
  "actions": [
    {
      "meta": {
        "arg": "item 1",
        "requestId": Any<String>,
        "requestStatus": "pending",
      },
      "payload": undefined,
      "type": "@todo/add/pending",
    },
    {
      "error": {
        "message": "Error get todo list.",
        "name": "Error",
        "stack": Any<String>,
      },
      "meta": {
        "aborted": false,
        "arg": "item 1",
        "condition": false,
        "rejectedWithValue": false,
        "requestId": Any<String>,
        "requestStatus": "rejected",
      },
      "payload": undefined,
      "type": "@todo/add/rejected",
    },
  ],
}
`;

exports[`ToDoSlice test addToDoList tests should success 1`] = `
{
  "actions": [
    {
      "meta": {
        "arg": "item 1",
        "requestId": Any<String>,
        "requestStatus": "pending",
      },
      "payload": undefined,
      "type": "@todo/add/pending",
    },
    {
      "meta": {
        "arg": "item 1",
        "requestId": Any<String>,
        "requestStatus": "fulfilled",
      },
      "payload": {
        "id": 100,
        "isFinish": false,
        "title": "item 1",
      },
      "type": "@todo/add/fulfilled",
    },
  ],
}
`;

exports[`ToDoSlice test editToDoItem tests should fail 1`] = `
{
  "actions": [
    {
      "meta": {
        "arg": {
          "id": 100,
          "title": "a",
        },
        "requestId": Any<String>,
        "requestStatus": "pending",
      },
      "payload": undefined,
      "type": "@todo/edit/pending",
    },
    {
      "error": {
        "message": "Error get todo list.",
        "name": "Error",
        "stack": Any<String>,
      },
      "meta": {
        "aborted": false,
        "arg": {
          "id": 100,
          "title": "a",
        },
        "condition": false,
        "rejectedWithValue": false,
        "requestId": Any<String>,
        "requestStatus": "rejected",
      },
      "payload": undefined,
      "type": "@todo/edit/rejected",
    },
  ],
}
`;

exports[`ToDoSlice test editToDoItem tests should success 1`] = `
{
  "actions": [
    {
      "meta": {
        "arg": {
          "id": 100,
          "title": "a",
        },
        "requestId": Any<String>,
        "requestStatus": "pending",
      },
      "payload": undefined,
      "type": "@todo/edit/pending",
    },
    {
      "payload": "edit",
      "type": "ToDo_List_Popup/closePopup",
    },
    {
      "payload": null,
      "type": "ToDo_List/setCurrent",
    },
    {
      "meta": {
        "arg": {
          "id": 100,
          "title": "a",
        },
        "requestId": Any<String>,
        "requestStatus": "fulfilled",
      },
      "payload": {
        "id": 100,
        "title": "a",
      },
      "type": "@todo/edit/fulfilled",
    },
  ],
}
`;

exports[`ToDoSlice test loadToDoList tests should fail 1`] = `
{
  "actions": [
    {
      "meta": {
        "arg": undefined,
        "requestId": Any<String>,
        "requestStatus": "pending",
      },
      "payload": undefined,
      "type": "@todo/list/load/pending",
    },
    {
      "error": {
        "message": "Error get todo list.",
        "name": "Error",
        "stack": Any<String>,
      },
      "meta": {
        "aborted": false,
        "arg": undefined,
        "condition": false,
        "rejectedWithValue": false,
        "requestId": Any<String>,
        "requestStatus": "rejected",
      },
      "payload": undefined,
      "type": "@todo/list/load/rejected",
    },
  ],
}
`;
